cmscontrib/ExportSubmissions.py

```diff
@@ -0,0 +1,213 @@
+    group.add_argument("--unique", action="store_true",
+                       help="if set, only the earliest best submission"
+                            " will be exported for each (user, task)")
+    group.add_argument("--best", action="store_true",
+                       help="if set, only the best submissions will be"
+                            " exported for each (user, task)")
+
+    args = parser.parse_args()
+
+    if not os.path.exists(args.output_dir):
+        os.mkdir(args.output_dir)
+    if not os.path.isdir(args.output_dir):
+        logger.critical("The output-dir parameter must point to a directory")
+        sys.exit(1)
+
+    with SessionGen() as session:
+        q = session.query(Submission)\
+            .join(Submission.task)\
+            .join(Submission.files)\
+            .join(Submission.results)\
+            .join(SubmissionResult.dataset)\
+            .join(Submission.participation)\
+            .join(Participation.user)\
+            .filter(Dataset.id == Task.active_dataset_id)\
+            .filter(SubmissionResult.score >= args.min_score)\
+            .with_entities(Submission.id, Submission.language, Submission.timestamp,
+                           SubmissionResult.score, File.filename, File.digest,
+                           User.id, User.username, User.first_name, User.last_name,
+                           Task.id, Task.name)
+
+        if args.contest_id:
+            q = q.filter(Participation.contest_id == args.contest_id)
+
+        if args.task_id:
+            q = q.filter(Submission.task_id == args.task_id)
+
+        if args.user_id:
+            q = q.filter(Participation.user_id == args.user_id)
+
+        if args.submission_id:
+            q = q.filter(Submission.id == args.submission_id)
+
+        results = q.all()
+
+        if args.unique or args.best:
+            results = filter_top_scoring(results, args.unique)
+
+        print("%s file(s) will be created." % len(results))
+        if raw_input("Continue? [Y/n] ").lower() not in ["y", ""]:
+            sys.exit(0)
+
+        done = 0
+        for row in results:
+            s_id, s_language, s_timestamp, sr_score, f_filename, f_digest, \
+                u_id, u_name, u_fname, u_lname, t_id, t_name = row
+
+            name = f_filename
+            if name.endswith(".%l"):
+                name = name[:-3]  # remove last 3 chars
+
+            filename = args.filename.format(id=s_id, name=name, ext=s_language,
+                                            time=s_timestamp, user=u_name)
+            filename = os.path.join(args.output_dir, filename)
+            if os.path.exists(filename):
+                logger.warning("Skipping file '%s' because it already exists",
+                               filename)
+
+            fso = FSObject.get_from_digest(f_digest, session)
+            assert fso is not None
+            with fso.get_lobject(mode="rb") as file_obj:
+                data = file_obj.read().decode('latin-1')
```

Comment:
Done. For some reason I remember that `decode('latin-1')` was necessary, but I just tested it now with `decode('utf-8')` and it works anyway, so...